framework/src/main/java/org/checkerframework/framework/type/SubtypeIsSubsetQualifierHierarchy.java

```diff
@@ -0,0 +1,134 @@
+     */
+    public SubtypeIsSubsetQualifierHierarchy(
+            Collection<Class<? extends Annotation>> qualifierClasses,
+            ProcessingEnvironment processingEnv) {
+        super(qualifierClasses, processingEnv.getElementUtils());
+        this.processingEnv = processingEnv;
+    }
+
+    @Override
+    protected boolean isSubtypeWithElements(
+            AnnotationMirror subAnno,
+            QualifierKind subKind,
+            AnnotationMirror superAnno,
+            QualifierKind superKind) {
+        if (subKind == superKind) {
+            List<String> superValues = extractValues(superAnno);
+            List<String> subValues = extractValues(subAnno);
+            return superValues.containsAll(subValues);
+        }
+        return subKind.isSubtypeOf(superKind);
+    }
+
+    @Override
+    protected AnnotationMirror leastUpperBoundWithElements(
+            AnnotationMirror a1,
+            QualifierKind qualifierKind1,
+            AnnotationMirror a2,
+            QualifierKind qualifierKind2,
+            QualifierKind lubKind) {
+        if (qualifierKind1 == qualifierKind2) {
+            List<String> a1Values = extractValues(a1);
+            List<String> a2Values = extractValues(a2);
+            LinkedHashSet<String> set = new LinkedHashSet<>(a1Values);
+            set.addAll(a2Values);
+            return createAnnotationMirrorWithValue(lubKind, set);
+        } else if (lubKind == qualifierKind1) {
+            return a1;
+        } else if (lubKind == qualifierKind2) {
+            return a2;
+        } else {
+            throw new BugInCF(
+                    "Unexpected QualifierKinds %s %s", qualifierKind1, qualifierKind2, lubKind);
+        }
+    }
+
+    @Override
+    protected AnnotationMirror greatestLowerBoundWithElements(
+            AnnotationMirror a1,
+            QualifierKind qualifierKind1,
+            AnnotationMirror a2,
+            QualifierKind qualifierKind2,
+            QualifierKind glbKind) {
+        if (qualifierKind1 == qualifierKind2) {
+            List<String> a1Values = extractValues(a1);
+            List<String> a2Values = extractValues(a2);
+            LinkedHashSet<String> set = new LinkedHashSet<>(a1Values);
+            set.retainAll(a2Values);
+            return createAnnotationMirrorWithValue(glbKind, set);
+        } else if (glbKind == qualifierKind1) {
+            return a1;
+        } else if (glbKind == qualifierKind2) {
+            return a2;
+        } else {
+            throw new BugInCF(
+                    "Unexpected QualifierKinds %s %s", qualifierKind1, qualifierKind2, glbKind);
+        }
+    }
+
+    /**
+     * Returns a mutable list containing the {@code values} element of the given annotation.
+     *
+     * @param anno an annotation
+     * @return a mutable list containing the {@code values} element; may be the empty list
+     */
+    private List<String> extractValues(AnnotationMirror anno) {
```

Comment:
This class could be changed to operate on lists of Objects rather than Strings. If you don't want to make that change, then the class comment should mention that the annotation element but be an array of strings.
\